org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -297,6 +302,12 @@ public void setSSLParameters(SSLParameters params) {
         if (parsed.getHostname() != null) {
             setHostname(parsed.getHostname());
         }
+
+        // When we have a non-zero number of ALPNs, use them in the
+        // negotiation.
+        if (parsed.getApplicationProtocols() != null) {
+            setApplicationProtocols(parsed.getApplicationProtocols());
+        }
```

Comment:
If we take a JDK8 `SSLParameters` (`params` here) and downcast to `JSSParameters` (`parsed`), we have no `getApplicationProtocols()` function in `params` so `parsed` will return `null`. However, because `SSLEngine` has a `setApplicationProtocols` method the following code would work:

```java
// SSLEngine inst;
inst.setApplicationProtocols(...);
inst.setSSLParameters(params);
```

If we overwrite it with `null` here, we could be overwriting intentional changes. We could try and detect this case but I'd prefer to leave it with the non-overwriting case to match all the other logic in the function anyways.

Reply:
In particular, if you take this source:

https://gist.github.com/cipherboy/f1144c36bb635257bd2b6bf2adb72513

If you compile and run with the default JDK (in my case 8), it'll fail to compile. That's fine since _this_ JDK8 doesn't have the method.

If you compile with JDK 11 and run with this older JDK8, you'll get a nice exception telling you that you can't do that because the method doesn't exist. 

If you compile with JDK11 and run with JDK11, it'll work fine. Presumably, this'd be the same behavior that we'd get when JDK8 updates. 

I think that's about as close to non-surprising behavior as you can get :-)

Reply:
Also to summarize some discussions on IRC:

BouncyCastle provides this in its BCSSLEngine/BCSSLParameters/... classes as well, for all JDK versions. When they compiled with JDK9, they'll add the methods to read the SSLParameters's value and put it into their BCSSLParameters copy.

Reply:
This probably why in general it's easier to bump the dependency (to Java 8u) such that certain feature is consistently available, especially when everybody is expected to use the latest Java 8 or later anyway.

If `SSLParameters.getApplicationProtocols()` is guaranteed to be available, the code could be as simple as this:
```
setApplicationProtocols(parsed.getApplicationProtocols());
```
If `SSLParameters.getApplicationProtocols()` is not consistently available, we probably should do something like this:
```
try {
    setApplicationProtocols(parsed.getApplicationProtocols());
} catch (NoSuchMethodException e) {
    // SSLParameters.getApplicationProtocols() is not available
}
```
Here we probably can implement a `JSSParameters.getApplicationProtocols()` that uses reflection to call `SSLParameters.getApplicationProtocols()`. If the method is not available, we simply let the caller handle the `NoSuchMethodException`.

Here the behavior would be more accurate: call `JSSEngine.setApplicationProtocols()` only if `SSLParameters.getApplicationProtocols()` is available.